lib/dry/initializer/rails.rb

```diff
@@ -9,6 +9,11 @@
   klass = model.is_a?(ActiveRecord::Relation) ? model.klass : model
 
   coercer = lambda do |value|
+    if value.is_a?(ActiveRecord::Base) && !value.instance_of?(klass)
```

Comment:
This check is fine if the `find_by` attribute is a default primary key. But what if a search is made by an associated record?

Suppose we have a `User` and an associated `User::State`.

```ruby
class User < ActiveRecord::Base
  has_one :state
end

class User::State < ActiveRecord::Base
  belongs_to: :user
end
```

And then we define somewhere:

```ruby
param :user, model: User, find_by: :state
```

This parameter allows to send a `user` either directly or by `state`. But you forbid this case because the state would case a failure as a model which is not a `User`.

That's why I find this check over-restricting

Reply:
That's a fair point. If I can find a way to allow the given classes and it's children, would that work for you?
My problem is that without this check, I can Pass `<Item#1>` to that service and it'll cast it to `<User#1>`, which doesn't seem right.

Reply:
I'd recommend going another way. Maybe it would be better if you added the special modifier like `strict: true` or whatever, which would enforce the behavior you need.

With such an option we could afford both cases. With `strict: true` it would allow only `model` class, without this option the check would be more permissive. And it would be 100% backward-compatible as well.

Reply:
Another interesting option here (I'm not sure though if it is as necessary as "interesting" )) is to check against the relation. For example, with a syntax:

```ruby
param :user, model: "User.active"
```

we could check that the model is not only of the desired class, but of the desired relation too.

Reply:
I'm going to implement the `strict: true` modifier.
For the option to check against a relation, it already exists in the gem right? You're just offering an alternative option?

Reply:
> Another interesting option here (I'm not sure though if it is as necessary as "interesting" )) is to check against the relation. For example, with a syntax:

Actually, it looks like overkill as for me :)
It's just an initializer (a bit advanced but still initializer) and probably should not have too heavy things.

Anyway, if this feature will land into the core some time, I'd suggest having syntax like this:

```ruby
param :user, model: User, scope: :active
param :user, model: User, scope: User.admins.active
```
Why?
- keep `model` to check against the type only
- use `scope` to check if the object's belonging to the scope, and raise a different error with the appropriate message
- having two options should help to keep gem's code simpler

---

> I'm going to implement the strict: true modifier.

I would be happy to have this options enabled by default. As we can see, regular users of this gem (like @nbcraft or guys from our @ergoserv team) facing the same problems because of too mild type validation. 

> This parameter allows to send a user either directly or by state. But you forbid this case because the state would case a failure as a model which is not a User.

That's a very tricky case. I'd prefer to have more straightforward behavior by default, with a possibility to make it milder consciously when is really needed.

---

@nepalez don't get me wrong - you are the owner and the boss :) I'm just sharing my thoughts. 
And many thanks to you for such a handy gem.

Reply:
> I would be happy to have this options enabled by default. As we can see, regular users of this gem (like @nbcraft or guys from our @ergoserv team) facing the same problems because of too mild type validation.

Well, I couldn't do this anyway for keeping backward compatibility with existing code.
But look, this gem is quite tiny and it is just a syntax sugar around `dry-initializer`. Maybe in your case it would be better to make a fork with redefined behavior for this check.

Reply:
Yes, sure. I just wanted to give some input to the official gem :)